### PR TITLE
feat: calculate visualisation popularity via the event log

### PR DESCRIPTION
**Pull request by niross**
### Description of change

Calculate visualisation catalogue item popularity...

- Popularity is average number of unique users per day in the past 28 days
- The past 28 days should be the midnight before the calculation is done, to the midnight 28 days before that
- For newly published datasets include only whole days they have been published in the calculation.
- Each user can be counted as a "viewer" of a given dataset at most once per day

### Checklist

* [x] Have tests been added to cover any changes?

### CODE REVIEW

File: dataworkspace/dataworkspace/apps/datasets/search.py [hunk 1]
```diff
@@ -450,15 +455,74 @@ def search_for_datasets(user, filters: SearchDatasetsFilters, matcher) -> tuple:
...
+    # The farthest we go back is 00:00 28 days ago
+    period_end = datetime.combine(datetime.utcnow(), time.min)
+    min_start_date = datetime.combine(datetime.utcnow(), time.min) - timedelta(days=28)
+
+    # If the vis was published < 28 days ago, start the count
+    # from the end of the day it was first published
+    published_date = datetime.combine(dataset.published_at, time.max)
+    period_start = published_date if published_date > min_start_date else min_start_date
+
+    return period_start, period_end
```
Comment (michalc) [inline dataworkspace/dataworkspace/apps/datasets/search.py:478]: Should these always be midnight UK time? (That is what I was thinking originally, but I now wonder if not worth it)

File: dataworkspace/dataworkspace/apps/datasets/search.py [hunk 2]
```diff
@@ -450,15 +455,74 @@ def search_for_datasets(user, filters: SearchDatasetsFilters, matcher) -> tuple:
...
+                EventLog.TYPE_VIEW_QUICKSIGHT_VISUALISATION,
+                EventLog.TYPE_VIEW_SUPERSET_VISUALISATION,
+                EventLog.TYPE_VIEW_VISUALISATION_TEMPLATE,
+            ],
+            timestamp__gt=period_start.replace(tzinfo=utc),
+            timestamp__lt=period_end.replace(tzinfo=utc),
+        )
+        .values("timestamp__date")
+        .annotate(user_count=Count("user", distinct=True))
+        .aggregate(total_users=Sum("user_count"))["total_users"]
```
Comment (michalc) [inline dataworkspace/dataworkspace/apps/datasets/search.py:508]: This is _much_ simpler than anything I came up with for this 👍

File: dataworkspace/dataworkspace/tests/datasets/test_search.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    with freeze_time(datetime.now() - timedelta(days=9)):
+        factories.EventLogFactory(
+            event_type=EventLog.TYPE_VIEW_SUPERSET_VISUALISATION,
+            related_object=vis,
+        )
+        factories.EventLogFactory(
+            event_type=EventLog.TYPE_VIEW_SUPERSET_VISUALISATION,
+            related_object=vis,
+        )
+    assert calculate_visualisation_average(vis) == 0.2222222222222222
```
Comment (michalc) [inline dataworkspace/dataworkspace/tests/datasets/test_search.py:104]: Great tests

**Review by michalc — APPROVED** [review]

File: dataworkspace/dataworkspace/apps/datasets/search.py [hunk 4]
```diff
@@ -450,15 +455,74 @@ def search_for_datasets(user, filters: SearchDatasetsFilters, matcher) -> tuple:
...
+    after it was published.
+    """
+    # The farthest we go back is 00:00 28 days ago
+    period_end = datetime.combine(datetime.utcnow(), time.min)
+    min_start_date = datetime.combine(datetime.utcnow(), time.min) - timedelta(days=28)
+
+    # If the vis was published < 28 days ago, start the count
+    # from the end of the day it was first published
+    published_date = datetime.combine(dataset.published_at, time.max)
+    period_start = published_date if published_date > min_start_date else min_start_date
```
[the remark this answers is not in the document]
Comment (niross) [reply dataworkspace/dataworkspace/apps/datasets/search.py:476]: Yes, definitely easier to read thanks

**Pull request merged**